### PR TITLE
Implement keyboard layout detection

### PR DESCRIPTION
**Pull request by jpnurmi**
The algorithm was borrowed from ubiquity.

A screencast with slightly outdated looks (see below for updated screenshots):

detect_keyboard_layout4

Close: #29

### CODE REVIEW

**Comment by jpnurmi** [comment]
Design update:
| 1 | 2 |
|---|---|
| Screenshot from 2021-08-24 16-28-44 | Screenshot from 2021-08-24 16-28-50 |

File: packages/ubuntu_desktop_installer/lib/pages/keyboard_layout/keyboard_layout_detector.dart [hunk 1]
```diff
@@ -0,0 +1,227 @@
+import 'dart:convert';
+
+/// Detects keyboard layout according to the steps defined in
+/// `/usr/share/console-setup/pc105.tree`.
```
Comment (robert-ancell) [inline packages/ubuntu_desktop_installer/lib/pages/keyboard_layout/keyboard_layout_detector.dart:4]: Nit: This class parses an arbitrary file, not necessarily the one specified here. The .tree file is installed as part of the installer, and the system one is not used anyway.

File: packages/ubuntu_desktop_installer/lib/pages/keyboard_layout/keyboard_layout_detector.dart [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+
+  /// The steps to take after the user has pressed the key.
+  final Map<int, int> steps;
+
+  /// The keycode user pressed;
+  late int _keycode;
+
+  // FIXME need to account for possible remapping. Find the API to translate
+  // kernel keycodes to X keycodes (xkb).
+  // MIN_KEYCODE = 8
```
Comment (robert-ancell) [inline packages/ubuntu_desktop_installer/lib/pages/keyboard_layout/keyboard_layout_detector.dart:91]: Is this last line of the comment supposed to be here? The following const seems to have replaced it.

File: packages/ubuntu_desktop_installer/lib/pages/keyboard_layout/keyboard_layout_detector.dart [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+          key = tokens[1];
+          break;
+        case 'FINDP': // FINDP <key>
+          // Equivalent to FIND, except that the user is asked to consider
+          // only the primary symbols (i.e. Plain and Shift).
+          assert(tokens.length == 2);
+          key = tokens[1];
+          break;
+        case 'YES':
+          // Direct the evaluating code to process step ## next if the
```
Comment (robert-ancell) [inline packages/ubuntu_desktop_installer/lib/pages/keyboard_layout/keyboard_layout_detector.dart:179]: Nit: Not sure what the double hash is referring to here. (same below).

**Review by robert-ancell — APPROVED** [review]
Just some small nits in the keyboard layout detector code (copy-pasted from an older project?) but otherwise looks great.

**Comment by mwhudson** [comment]
Did you realize subiquity parses pc105.tree server side and has an API?

```
(for-20.04.3)mwhudson@anduril:~/src/subiquity$ curl --silent --unix-socket .subiquity/socket a/keyboard/steps?index='"1"' | jq 
{
  "symbols": [
    "z"
  ],
  "keycodes": [
    [
      21,
      "2"
    ],
    [
      44,
      "3"
    ]
  ],
  "$type": "StepPressKey"
}
```

**Comment by jpnurmi** [comment]
@mwhudson Symbols are ok but keycodes are somehow twisted in the data provided by subiquity.

For comparison, step 0 should be: [9,1], [5,43], [43,57], [13,69]...

`$ less /usr/share/console-setup/pc105.tree`

```
STEP 0
PRESS *
PRESS ¤
PRESS n
PRESS u
PRESS v
PRESS y
PRESS γ
PRESS ν
PRESS υ
PRESS г
PRESS у
PRESS ה
CODE 9 1
CODE 5 43
CODE 43 57
CODE 13 69
CODE 27 73
CODE 53 82
CODE 4 84
CODE 26 85
CODE 6 38
CODE 41 86
CODE 12 87
CODE 21 88
CODE 22 113
CODE 47 113
CODE 49 113
CODE 44 115
CODE 20 41
CODE 33 41
CODE 38 41
CODE 52 41
CODE 17 31
CODE 35 31
CODE 18 21
CODE 23 86
CODE 30 86
CODE 39 86
CODE 46 86
CODE 34 42
STEP 1
[...]
```
vs.

`$ curl --silent --unix-socket .subiquity/socket a/keyboard/steps?index='"0"' | jq`

```json
{
  "symbols": [
    "*",
    "¤",
    "n",
    "u",
    "v",
    "y",
    "γ",
    "ν",
    "υ",
    "г",
    "у",
    "ה"
  ],
  "keycodes": [
    [
      5,
      "1"
    ],
    [
      43,
      "16"
    ],
    [
      13,
      "23"
    ],
    [
      9,
      "32"
    ],
    [
      53,
      "73"
    ],
    [
      41,
      "75"
    ],
    [
      27,
      "76"
    ],
    [
      12,
      "85"
    ],
    [
      4,
      "86"
    ],
    [
      6,
      "69"
    ],
    [
      26,
      "87"
    ],
    [
      21,
      "88"
    ],
    [
      22,
      "113"
    ],
    [
      47,
      "113"
    ],
    [
      49,
      "113"
    ],
    [
      44,
      "115"
    ],
    [
      23,
      "75"
    ],
    [
      30,
      "75"
    ],
    [
      39,
      "75"
    ],
    [
      46,
      "75"
    ],
    [
      17,
      "62"
    ],
    [
      35,
      "62"
    ],
    [
      34,
      "72"
    ],
    [
      18,
      "52"
    ],
    [
      20,
      "71"
    ],
    [
      33,
      "71"
    ],
    [
      38,
      "71"
    ],
    [
      52,
      "71"
    ]
  ],
  "$type": "StepPressKey"
}
```

**Comment by mwhudson** [comment]
On Fri, 27 Aug 2021 at 01:11, J-P Nurmi ***@***.***> wrote:

> @mwhudson <https://github.com/mwhudson> Could the list of StepPressKey
> keycodes be a map/object for easier lookup and sound types? In Dart, Map<int,
> String> vs. List<List<dynamic>>.
>
They are intended to be interpreted that way but you can't do an
int->string mapping in JSON (this is one area where subiquity's API's
serialization is not self-documenting).

**Comment by mwhudson** [comment]
On Thu, 26 Aug 2021 at 23:54, J-P Nurmi ***@***.***> wrote:

> Thanks for investigating. I guess the latest and greatest is better. I was
> just mostly confused why the keys were different.
>
Yeah I was surprised too! It led me to discover that pc105.tree is
generated by a script from other keyboard data files at package build time
(and that this script is added in Ubuntu delta -- I guess d-i in Debian
doesn't do keyboard detection like this). So perhaps that keyboard data
changed to cause the script to build a new tree or it's not deterministic
about step numbering. I guess any day you learn something is a good day,
right? :)

**Comment by seb128** [comment]
Testing the changes the feature seems to work from an user perspective, it correctly detects an azerty layout asking the same steps as ubiquity is doing

**Comment by jpnurmi** [comment]
Thanks for testing!

**Pull request merged**